### PR TITLE
Pathfinding Implementation

### LINKED ISSUE #78: Implement pathfinding
Implement A* pathfinding for use with ships in the MVP. The pathfinding algorithm can be bound to the grid structure of the game board.

### PR DESCRIPTION
**Pull request by JoXW100**
Implementation of the A* Algorithm to find paths in a scene. 
- It can avoid other objects in the scene at given depth orders.
`gameObject.test` has been modified to test the new feature. 
- It can handle non grid based scenes.

Fixes #78 

image

### CODE REVIEW

**Comment by joelsiks** [comment]
Rename the `PF_Node` to `PFNode`.

File: src/engine/PathFinding.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+#include "PathFinding.hpp"
+#include <set>
```
Comment (joelsiks) [inline src/engine/PathFinding.cpp:2]: Standard library includes should be before project includes.

File: src/engine/PathFinding.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        !IsValidPosition(dest, Vector2(__FLT_EPSILON__), quadTree,
+                         checkedOrders)) {
+        return std::deque<Vector2>();
+    }
+
+    const size_t width =
+        static_cast<size_t>((bounds.Width() / detailLevel) + 1);
+    const size_t height =
+        static_cast<size_t>((bounds.Height() / detailLevel) + 1);
+    // world grid
```
Comment (joelsiks) [inline src/engine/PathFinding.cpp:25]: world?

File: src/engine/PathFinding.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    Node *grid = new Node[width * height];
+
+    const size_t startIndex =
+        ConvertVectorToNodeIndex(start, width, detailLevel);
+    const size_t endIndex = ConvertVectorToNodeIndex(dest, width, detailLevel);
+
+    std::set<size_t, PathFinding::Comparator> queue =
```
Comment (joelsiks) [inline src/engine/PathFinding.cpp:32]: Should probably have a more descriptive name instead of just queue. A queue of what?

Comment (joelsiks) [reply src/engine/PathFinding.cpp:32]: Position/Index/...?

File: src/engine/PathFinding.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        std::set<size_t, PathFinding::Comparator>(
+            PathFinding::Comparator(grid));
+    queue.insert(startIndex);
+    grid[startIndex].visited = true;
+    grid[startIndex].path = NULLVALUE;
+
+    size_t neighbors[8];
+    while (!queue.empty()) {
+        const size_t nodeIndex = *queue.begin();
+        queue.erase(queue.begin());
```
Comment (joelsiks) [inline src/engine/PathFinding.cpp:42]: `queue.pop()` would be nicer

File: src/engine/PathFinding.hpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+namespace admirals {
+
+class PathFinding {
+private:
+    struct Node {
+        bool visited = false;
+        float g = 0;
+        float h = 0;
+        float f = 0;
```
Comment (joelsiks) [inline src/engine/PathFinding.hpp:19]: Either rename these to more descriptive values or add comments to them

File: src/engine/PathFinding.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            if (checkedOrders.contains(o->order()) &&
+                bounds.Overlaps(o->GetBoundingBox())) {
+                return false;
+            }
+        }
+        return true;
+    }
+    /// @description Finds the step the car should take (since traffic is always
+    /// moving we do not need to chain the whole path as we need to do the
+    /// algorithm from this new node)
```
Comment (joelsiks) [inline src/engine/PathFinding.hpp:82]: Car?

File: src/engine/QuadTree.cpp [hunk 7]
```diff
@@ -74,6 +74,38 @@ QuadTree::GetObjectsAtPosition(const Vector2 &position) const {
     return prev->data;
 }
 
+std::unordered_set<std::shared_ptr<IDisplayable>>
+QuadTree::GetObjectsAtLocation(const Rect &location) const {
```
Comment (joelsiks) [inline src/engine/QuadTree.cpp:78]: I mean this is essentially "GetObjectsInsideRect"...

**Review by joelsiks — COMMENTED** [review]
See comments.

File: src/engine/QuadTree.cpp (hunk 7 above)
Comment (joelsiks) [reply src/engine/QuadTree.cpp:78]: Maybe "GetObjectsInsideArea"

File: src/engine/PathFinding.cpp (hunk 4 above)
Comment (JoXW100) [reply src/engine/PathFinding.cpp]: Yes, though std::set does not support that.

File: src/engine/PathFinding.hpp (hunk 5 above)
Comment (JoXW100) [reply src/engine/PathFinding.hpp]: The A* algorithm refers to these values as g, h, f. They do not have clear names (except h for heuristc).

**Review by OskarG0 — APPROVED** [review]

**Review by joelsiks — APPROVED** [review]
Other han `orderedNodeIndexQueue` not being an actual queue, but a set, this looks good.

**Pull request merged**